File: internal/log/json.go
package log

import (
	"encoding/json"
	"fmt"
	"os"
)

type JsonAction struct {
	Action string `json:"action"`
	Path   string `json:"path"`
}

type JsonFile struct {
	Status string `json:"status"`
}

type JsonIssue struct {
	Severity string `json:"severity"`
	Message  string `json:"message"`
	Location string `json:"path"`
}

type JsonLog struct {
	Actions []JsonAction `json:"actions"`
	Issues  []JsonIssue  `json:"issues"`
}

var JsonLogger *JsonLog = &JsonLog{}
var JsonLogging bool = false

func Dump(s interface{}) error {
	if !JsonLogging {
		return nil
	}

	jsonBytes, err := json.MarshalIndent(s, "", "    ")
	if err != nil {
		return err
	}

	os.Stdout.Write(jsonBytes)
	fmt.Print("\n")

	return nil
}

func DumpAndExit(code int) {
	err := Dump(JsonLogger)
	if err != nil {
		fmt.Println("Failed to dump JSON log:", err)
	}

	os.Exit(code)
}

File: src/internal/log/json.rs
use serde::{Deserialize, Serialize};
use serde_json::Result;

#[derive(Serialize, Deserialize)]
pub struct JsonAction {
    pub action: String,
    pub path: String
}

#[derive(Serialize, Deserialize)]
pub struct JsonFile {
    pub status: String
}

#[derive(Serialize, Deserialize)]
pub struct JsonIssue {
    pub severity: String,
    pub message: String,
    pub location: String
}

#[derive(Serialize, Deserialize)]
pub struct JsonLog {
    pub actions: Vec<JsonAction>,
    pub issues: Vec<JsonIssue>
}

#[derive(Serialize, Deserialize)]
pub struct JsonLogger(Vec<JsonLog>);

